lib/sqlalchemy/sql/schema.py

```diff
@@ -5664,3 +5707,239 @@ def _copy(self, **kw: Any) -> Identity:
+        self.end = coercions.expect(roles.DDLExpressionRole, end)
+        self.system = system
+        self.primary_key = primary_key
+
+    @property
+    def key(self) -> str:
+        return self.name
+
+    def _get_and_validate_column(
+        self, table: Table, column: Union[str, Column]
+    ) -> Column:
+        """Accept a string or Column object, validate it is in metadata and
+        return the Table object."""
+        if isinstance(column, Column):
+            if column.table is not table:
+                raise exc.ArgumentError(
+                    "Given column object '%s' does not belong to Table '%s'"
+                    % (column.name, table.description)
+                )
+            return column
+
+        col = table.c.get(str(column))
+
+        if col is None:
+            raise exc.ArgumentError(
+                "Cannot find column '%s' in Table '%s'"
+                % (column, table.description)
+            )
+        return col
+
+    def _set_parent(self, table: Table, **kw: Any) -> None:
+        existing = getattr(self, "table", None)
+        if existing is not None and existing is not table:
+            raise exc.ArgumentError(
+                "Period object '%s' already assigned to Table '%s'"
+                % (self.name, existing.description)
+            )
+        # Update parameters to Column type
+        if not self.system:
+            self.start = self._get_and_validate_column(table, self.start)
+            self.end = self._get_and_validate_column(table, self.end)
+
+        if self.primary_key:
+            table.primary_key._replace(self)
+        elif self.key in table.primary_key:
+            raise exc.ArgumentError(
+                "Trying to redefine primary-key period '%s' as a "
+                "non-primary-key column on table '%s'"
+                % (self.name, table.fullname)
+            )
+
+        self.table = table
+
+        table.periods.add(self, self.name)
+
+    def _copy(self, target_table: Table = None, **kw):
+        p = Period(name=self.name, start=str(self.start), end=str(self.end))
+        return self._schema_item_copy(p)
+
+
+class ApplicationTimePeriod(Period):
+    """:class:`_schema.ApplicationTimePeriod` is an alias for
+    :class:`_schema.Period` and accepts the same constructor.
+
+    Both :class:`_schema.ApplicationTimePeriod` and :class:`_schema.Period`
+    define a period construction, i.e. "PERIOD FOR" syntax (which is used for
+    application versioning) and may be used interchangeably, based on what is
+    most clear for the given application.
+
+    See the linked documentation below for complete details.
+
+    .. versionadded:: TODO
+    """
+
+
+class SystemTimePeriod(Period):
+    """Sets ``WITH SYSTEM VERSIONING`` on a table ant optionally configures a
+    ``SYSTEM_TIME`` period.
+
+    If a backend needs an alternative system time period name instead of
+    ``SYSTEM_TIME``, set it by subclassing this class and overriding the
```

Comment:
**Federico Caselli** (CaselIT) wrote:

these are not correct, str(self.start/end) may not generated the key of the column.  table does not exist.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803